exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/metrics_exporter/__init__.py

```diff
@@ -0,0 +1,191 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""OTLP Metrics Exporter"""
+
+import logging
+
+from opentelemetry.exporter.otlp.exporter import (
+    OTLPExporterMixin,
+    _get_resource_data,
+)
+from opentelemetry.proto.collector.metrics.v1.metrics_service_pb2 import (
+    ExportMetricsServiceRequest,
+)
+from opentelemetry.proto.collector.metrics.v1.metrics_service_pb2_grpc import (
+    MetricsServiceStub,
+)
+from opentelemetry.proto.common.v1.common_pb2 import StringKeyValue
+from opentelemetry.proto.metrics.v1.metrics_pb2 import (
+    DoubleDataPoint,
+    InstrumentationLibraryMetrics,
+    Int64DataPoint,
+)
+from opentelemetry.proto.metrics.v1.metrics_pb2 import (
+    Metric as CollectorMetric,
+)
+from opentelemetry.proto.metrics.v1.metrics_pb2 import (
+    MetricDescriptor,
+    ResourceMetrics,
+)
+from opentelemetry.sdk.metrics import (
+    Counter,
+    SumObserver,
+    UpDownCounter,
+    UpDownSumObserver,
+    ValueObserver,
+    ValueRecorder,
+)
+from opentelemetry.sdk.metrics.export import (
+    MetricsExporter,
+    MetricsExportResult,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def _get_data_points(sdk_metric, data_point_class):
+
+    data_points = []
+
+    for (
+        label,
+        bound_counter,
+    ) in sdk_metric.instrument.bound_instruments.items():
+
+        string_key_values = []
+
+        for label_key, label_value in label:
+            string_key_values.append(
+                StringKeyValue(key=label_key, value=label_value)
+            )
+
+        for view_data in bound_counter.view_datas:
+
+            if view_data.labels == label:
+
+                data_points.append(
+                    data_point_class(
+                        labels=string_key_values,
+                        value=view_data.aggregator.current,
+                    )
+                )
+                break
+
+    return data_points
+
+
+def _get_temporality(instrument):
+    # pylint: disable=no-member
+    if isinstance(instrument, (Counter, UpDownCounter)):
+        temporality = MetricDescriptor.Temporality.DELTA
```

Comment:
Yes temporality should be based off statefulness. Currently it is a feature of the batcher but this might be changed soon.

Reply:
I believe OTLP is supposed to describe the metrics on the wire rather than the instruments that created them, so that whatever is receiving the OTLP can make sense of it. For instance, if the collector was receiving `DELTA`s and then was supposed to export them to prometheus (which doesn't accept deltas for a cumulative metric), it knows to convert the `DELTA`s to `CUMULATIVE`s before exporting to prometheus.

So in the example above, if the SDK sends `10, 20` to the collector over OTLP with `DELTA`, then the collector converts them to cumulatives `10, 30` which is sending incorrect values to prometheus because it was already converted to cumulative in the SDK. That's why I think we should choose based on `stateful`.

- The metrics API spec doesn't mention "stateful" at all. I think that's why the table there says that Counter/UpDownCounter would be delta, because it is assuming the SDK just sends off the DELTA is sums for the interval (`stateful=False`)
- [OTEP #131](https://github.com/open-telemetry/oteps/pull/131/files#diff-4f2740c98df4160f9ea8511dffc820c4) (was just merged) changes the default behavior of OTLP exporters to send `CUMULATIVES` for Counter/UpDownCounter and allows configuration. The python sdk's "stateful" seems like that configuration. Point being, I don't think it's imperative to follow the table in the metrics API spec.

I think we will need to revisit this again very soon since OTLP is changing quickly, so I wouldn't worry too much about it not being perfect yet.